### PR TITLE
feat: Support for committing externally produced root manifests

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
## What changes are proposed in this pull request?

Adds `Transaction::with_external_root_manifest` (+ FFI) to commit a caller-produced root
manifest as a table's content root, gated on `adaptiveMetadata-preview`. Kernel validates the
file and any existing checkpoint, then writes a `CheckpointAction` pointing at it. 

## How was this change tested?

Unit tests for checkpoint action construction and `scan_non_content_metadata`'s scan/CRC
paths. Integration tests against a real adaptiveMetadata-preview table covering the happy
path, rejections, and domain metadata/txn merging. FFI has error and success path tests.

Follow-ups:
- Replacing an existing checkpoint is only unit-tested, no integration test yet.
- No sidecar spill for domain metadata/txns.
- Caching the checkpoint-action lookup, separate PR.

### CODE REVIEW

File: ffi/src/transaction/mod.rs [hunk 1]
```diff
@@ -2184,6 +2222,90 @@ mod tests {
...
+        .await?;
+
+        let engine = engine_handle_for_store(Arc::clone(&store));
+        let table_path = table_url.to_string();
+        let table_path_str = table_path.as_str();
+        let txn = ok_or_panic(unsafe {
+            transaction(kernel_string_slice!(table_path_str), engine.shallow_copy())
+        });
+
+        let manifest_path = table_url.join("metadata/root-v1.parquet")?.to_string();
```
Comment (emkornfield) [inline ffi/src/transaction/mod.rs:2276]: maybe make a helper here to avoid boiler plate in the two tests?

File: ffi/src/transaction/mod.rs [hunk 2]
```diff
@@ -2184,6 +2222,90 @@ mod tests {
...
+    async fn test_with_external_root_manifest_requires_the_feature(
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let (store, _test_engine, table_url) =
+            test_utils::engine_store_setup("test_external_root_manifest", None);
+        let (table_path, engine, builder) = create_table_builder(
+            &store,
+            &table_url,
+            vec![StructField::nullable("id", DataType::INTEGER)],
+        );
+        build_and_commit(builder, &engine);
```
Comment (emkornfield) [inline ffi/src/transaction/mod.rs]: assert somewhere that the necessary feature is not enabled, in case we start enabling it by default at some point.

File: kernel/src/snapshot/mod.rs [hunk 3]
```diff
@@ -745,6 +756,82 @@ impl Snapshot {
...
+        &self,
+        engine: &dyn Engine,
+    ) -> DeltaResult<(
+        Vec<DomainMetadata>,
+        Vec<SetTransaction>,
+        Option<CheckpointAction>,
+    )> {
+        // Domain metadata/set transactions each come from exactly one source: an at-version CRC
+        // (`Complete`), or a log scan. Modeling this as an enum rather than two correlated
+        // `Option`s makes that mutual exclusion a type-level fact instead of a runtime invariant.
```
Comment (emkornfield) [inline kernel/src/snapshot/mod.rs]: borderline AI slop?

File: kernel/src/transaction/external_root_manifest.rs [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+            read_snapshot,
+        })
+    }
+
+    /// Builds the checkpoint action referencing this commit's file at `commit_version`, carrying
+    /// `protocol`, `metadata`, `domain_metadata`, and `transactions`.
+    ///
+    /// Errors if `existing_checkpoint_action` doesn't cover `read_snapshot_version`.
+    #[allow(clippy::too_many_arguments)]
+    pub(super) fn checkpoint_action(
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: does this method belong in actions/mod.rs?

File: kernel/src/transaction/external_root_manifest.rs [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+        &self,
+        existing_checkpoint_action: Option<&CheckpointAction>,
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: document how this checkpointaction is reused in the contract?

File: kernel/src/transaction/external_root_manifest.rs [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+            file,
+            read_snapshot,
+        })
+    }
+
+    /// Builds the checkpoint action referencing this commit's file at `commit_version`, carrying
+    /// `protocol`, `metadata`, `domain_metadata`, and `transactions`.
+    ///
+    /// Errors if `existing_checkpoint_action` doesn't cover `read_snapshot_version`.
+    #[allow(clippy::too_many_arguments)]
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: Still reading but maybe this would affect method placement but should this take a Snapshot instead to avoid too many arguments?

File: kernel/src/snapshot/mod.rs (hunk 3 above)
Comment (amogh-jahagirdar) [reply kernel/src/snapshot/mod.rs]: Yeah this definitley is AI slop :( will clean this up

File: ffi/src/transaction/mod.rs (hunk 1 above)
Comment (kyli87) [reply ffi/src/transaction/mod.rs:2276]: +1 - or you could parameterize it with rstest and a simple bool like `feature_enabled`

File: kernel/src/transaction/external_root_manifest.rs [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+    CheckpointAction, ContentRoot, DomainMetadata, Metadata, Protocol, SetTransaction,
+};
+use crate::error::Error;
+use crate::snapshot::SnapshotRef;
+use crate::table_features::TableFeature;
+use crate::utils::require;
+use crate::{version_as_i64, DeltaResult, FileMeta, Version};
+
+/// A caller-supplied root manifest committed as the table's content root.
+pub(super) struct ExternalRootManifest {
```
Comment (kyli87) [inline kernel/src/transaction/external_root_manifest.rs]: Naming suggestion (for throughout the PR): 
- I feel like "External" is an unnecessary distinction and may cause more confusion. From kernel Txn perspective, you're just providing a root manifest to commit as part of checkpoint action. We don't preface other connector provided values with "External".
- I personally prefer naming that is more obviously related to the field name that it eventually populates e.g `contentRoot` on the `checkpoint` action.

My suggestion would be to name this struct just `RootManifest` and on the Txn methods, maybe `with_checkpoint_root` or `with_checkpoint_content_root`?

File: kernel/src/snapshot/mod.rs [hunk 8]
```diff
@@ -745,6 +756,82 @@ impl Snapshot {
             .collect())
     }
 
+    /// Domain metadata, set transactions, and the latest checkpoint action on this snapshot.
+    #[cfg(feature = "adaptive-metadata-in-dev")]
+    pub(crate) fn scan_non_content_metadata(
```
Comment (kyli87) [inline kernel/src/snapshot/mod.rs]: This doesn't seem like a general purpose helper that belongs on Snapshot. Could we move this logic instead into `ExternalRootManifest` to encapsulate all of the logic for building the `checkpoint` action? 

log_segment is exposed as internal_api so there shouldn't be any compiler visibility restrictions.

File: kernel/src/snapshot/mod.rs (hunk 3 above)
Comment (kyli87) [reply kernel/src/snapshot/mod.rs]: +1 - also the enum design seems unnecessary, AI loves to insert one-off enums. I would just:
1. Initialize Optional domain_metadata, set_transaction, etc variables
2. First try CRC
3. Then try scan if any of the variables are still None.

File: kernel/src/transaction/external_root_manifest.rs (hunk 6 above)
Comment (kyli87) [reply kernel/src/transaction/external_root_manifest.rs]: +1, I would try to centralize all of the "checkpoint action" building logic internally in one helper like `compute_checkpoint_action` instead of computing it in one place and then passing it to another.

File: kernel/src/transaction/mod.rs [hunk 9]
```diff
@@ -487,13 +500,64 @@ impl<S> Transaction<S> {
         let remove_actions =
             self.generate_remove_actions(engine, self.remove_files_metadata.iter(), &[])?;
 
+        // Step 6b: checkpoint action for an external root manifest commit, if configured.
+        #[cfg(feature = "adaptive-metadata-in-dev")]
+        let checkpoint_action = self
```
Comment (kyli87) [inline kernel/src/transaction/mod.rs]: Related to prior comments - let's centralize this logic somewhere. At a minimum, in a separate helper like `generate_checkpoint_action` following the pattern above for other action types

File: kernel/src/transaction/mod.rs [hunk 10]
```diff
@@ -789,6 +853,19 @@ impl<S> Transaction<S> {
         Ok(())
     }
 
+    /// Validate that an external root manifest commit has no file actions.
+    #[cfg(feature = "adaptive-metadata-in-dev")]
+    fn validate_external_root_manifest_semantics(&self) -> DeltaResult<()> {
```
Comment (kyli87) [inline kernel/src/transaction/mod.rs]: We should centralize all of the validation logic here e.g checking feature enablement

**Review by kyli87 — COMMENTED** [review]
First pass - some comments on code structure.

File: kernel/src/transaction/external_root_manifest.rs (hunk 7 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/external_root_manifest.rs]: Well I think the issue is at some point we'll probably have a `RootManifest` or `ContentRoot` which is an inmemory model of the entries in that root so it'll likely conflict with whatever naming we do then. Whereas here we make that distinction intentionally because we're just trying to handle a pointer to the _on-disk_ root that some other system produced.

File: kernel/src/transaction/external_root_manifest.rs [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+    fn dummy_snapshot() -> DeltaResult<SnapshotRef> {
+        let engine = SyncEngine::new_with_store(Arc::new(InMemory::new()));
+        let schema = schema_ref! { nullable "id": INTEGER };
+        let _ = create_table("memory:///", schema, "test")
+            .build(&engine, Box::new(FileSystemCommitter::new()))?
+            .commit(&engine)?;
+        Snapshot::builder_for("memory:///").build(&engine)
+    }
+
+    fn mock_checkpoint_action(path: &str, version: Version) -> DeltaResult<CheckpointAction> {
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: nit: is mock used in the code base?  This seems more like a fake or a minimal checkpoint action.

File: kernel/src/transaction/external_root_manifest.rs [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+            .build(&engine, Box::new(FileSystemCommitter::new()))?
+            .commit(&engine)?;
+        let table_root = Snapshot::builder_for("memory:///")
+            .build(&engine)?
+            .table_root()
+            .clone();
+
+        let domain_metadata = DomainMetadata::new("test.domain".to_string(), "{}".to_string());
+        write_commit(
+            &engine,
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: nit: maybe make a closure over engine and table_root to avoid boiler plate?

File: kernel/src/transaction/external_root_manifest.rs [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+        assert_eq!(transactions[0].app_id, "app-1");
+        assert_eq!(existing_checkpoint.map(|c| c.version()), Some(1));
+        Ok(())
+    }
+
+    #[test]
+    fn scan_non_content_metadata_uses_crc_fast_path() -> DeltaResult<()> {
+        let engine = SyncEngine::new_with_store(Arc::new(InMemory::new()));
+        let schema = schema_ref! { nullable "id": INTEGER };
+        let _ = create_table("memory:///", schema, "test")
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: move table creatieon to helper?

File: kernel/src/transaction/external_root_manifest.rs [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+        let content_root = ContentRoot::new(self.file.location.to_string(), size_in_bytes, version);
+
+        Ok(CheckpointAction {
+            version,
+            content_root,
+            protocol,
+            metadata,
+            transactions,
+            domain_metadata,
+            txn_sidecars: vec![],
```
Comment (emkornfield) [inline kernel/src/transaction/external_root_manifest.rs]: we should probably have a TODO here or something to figure out how to integrate with side-cars once they exist?

File: kernel/src/transaction/update.rs [hunk 15]
```diff
@@ -148,6 +154,17 @@ impl Transaction {
         self
     }
 
+    /// Commits `file` as the table's root manifest.
+    #[cfg(feature = "adaptive-metadata-in-dev")]
+    pub fn with_external_root_manifest(mut self, file: FileMeta) -> DeltaResult<Self> {
+        let read_snapshot = self.read_snapshot_opt.clone().ok_or_else(|| {
+            Error::internal_error("read_snapshot() called on create-table transaction")
```
Comment (emkornfield) [inline kernel/src/transaction/update.rs:162]: is this the right error, we are inside the update method module

File: kernel/tests/integration/write/external_root_manifest.rs [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+const WRITER_FEATURES: &[&str] = &[
+    "columnMapping",
+    "deletionVectors",
+    "rowTracking",
+    "domainMetadata",
+    "inCommitTimestamp",
+    "adaptiveMetadata-preview",
+];
+
+/// Reads the raw actions of type `action_type` (e.g. `"checkpoint"`) from a commit file.
```
Comment (emkornfield) [inline kernel/tests/integration/write/external_root_manifest.rs]: there isn't an existing helper someplace for this?

File: kernel/tests/integration/write/external_root_manifest.rs [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+    txn.commit(&engine)?.unwrap_committed();
+
+    let checkpoint_actions =
+        read_commit_actions(store.as_ref(), &table_url, 2, "checkpoint").await?;
+    assert_eq!(checkpoint_actions.len(), 1);
+    let entries = checkpoint_actions[0]
+        .as_array()
+        .expect("checkpoint is an array");
+
+    let domain_metadata: std::collections::HashMap<String, String> = entries
```
Comment (emkornfield) [inline kernel/tests/integration/write/external_root_manifest.rs]: import instead of full reference?

**Review by emkornfield — COMMENTED** [review]
logic seems reasonable, seems like we could make tests tighter and some better code organization.

File: kernel/src/transaction/external_root_manifest.rs (hunk 7 above)
Comment (kyli87) [reply kernel/src/transaction/external_root_manifest.rs]: Then maybe `RootManifestFile` (indicating its just a pointer to a file)? And then inmemory struct can be called `RootManifest` still. I just feel like "external" is vague

Comment (amogh-jahagirdar) [reply kernel/src/transaction/external_root_manifest.rs]: Good points, RootManifestFile sounds reasonable to me. That is a shared concept that can be used for both manifests written by another client and for what kernel will produce as well.

File: kernel/src/transaction/mod.rs (hunk 10 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/mod.rs]: I did add the feature enablement to this validation, but I think some of the root manifest validation belongs in the constructor. Largely drawing a distinction between "Is this a valid root manifest (paths/qualities of the URI etc) in the context of kernel" vs "Is it valid to even produce a root manifest in a transaction"

File: kernel/src/transaction/mod.rs (hunk 9 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/mod.rs]: Done!

File: kernel/src/snapshot/mod.rs (hunk 8 above)
Comment (amogh-jahagirdar) [reply kernel/src/snapshot/mod.rs]: Fair point, it's not general purpose at the moment, it's really just for getting content of non content metadata just for the checkpoint. So we can keep it out of snapshot.

File: kernel/src/transaction/external_root_manifest.rs (hunk 6 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/external_root_manifest.rs]: Done! I was prompting claude that I wanted to pass very specific arguments to functions to make it obvious what's actually needed just from a signature but this is at the point of too many args.

File: kernel/tests/integration/write/root_manifest_file.rs [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+    let content_root = entries
+        .iter()
+        .find_map(|e| e.get("contentRoot"))
+        .expect("contentRoot entry");
+    assert_eq!(content_root["path"], json!(file.location.to_string()));
+    assert_eq!(content_root["sizeInBytes"], json!(1024));
+
+    assert!(entries.iter().any(|e| e.get("protocol").is_some()));
+    assert!(entries.iter().any(|e| e.get("metaData").is_some()));
```
Comment (kyli87) [inline kernel/tests/integration/write/root_manifest_file.rs]: Also validate that its the correct P&M?

File: kernel/src/transaction/root_manifest_file.rs [hunk 19]
```diff
@@ -0,0 +1,586 @@
...
+use crate::crc::{merge_domain_metadata, DomainMetadataState, SetTransactionState};
+use crate::error::Error;
+use crate::log_segment::DomainMetadataMap;
+use crate::schema::StructType;
+use crate::snapshot::SnapshotRef;
+use crate::table_configuration::TableConfiguration;
+use crate::utils::require;
+use crate::{version_as_i64, DeltaResult, Engine, FileMeta, RowVisitor as _, Version};
+
+/// A pointer to an on-disk root manifest file, produced by some other system, to be committed as
```
Comment (kyli87) [inline kernel/src/transaction/root_manifest_file.rs]: nit: "produced by some other system," is an unnecessary distinction?

File: kernel/src/transaction/root_manifest_file.rs [hunk 20]
```diff
@@ -0,0 +1,586 @@
...
+            require!(
+                existing.version() >= read_snapshot_version,
+                Error::generic(format!(
+                    "root manifest file commit requires no delta log commits pending replay since \
+                     the last checkpoint; existing checkpoint covers version {} but snapshot is \
+                     at {read_snapshot_version}",
+                    existing.version()
+                ))
+            );
+        }
```
Comment (kyli87) [inline kernel/src/transaction/root_manifest_file.rs:92]: I'm a bit confused by this check - if we found an existing checkpoint, why would it have to be `>= read_snapshot_version`? Aren't we trying to create a new checkpointAction here, so isn't it valid that prior checkpoint would be older?

File: kernel/src/transaction/root_manifest_file.rs [hunk 21]
```diff
@@ -0,0 +1,586 @@
...
+        );
+        transactions.extend(
+            set_transactions
+                .iter()
+                .cloned()
+                .map(|txn| (txn.app_id.clone(), txn)),
+        );
+        let expiration_timestamp =
+            calculate_transaction_expiration_timestamp(table_config.table_properties())?;
+        transactions.retain(|_, txn| !txn.is_expired(expiration_timestamp));
```
Comment (kyli87) [inline kernel/src/transaction/root_manifest_file.rs:79]: When we merge these, I think we need to only keep the most recent per appId? I assume `scan_non_content_metadata` will only return the most recent per appId but then this merging logic needs to do the same if we overwrote an appId?

File: kernel/src/transaction/root_manifest_file.rs [hunk 22]
```diff
@@ -0,0 +1,586 @@
...
+        let schema = StructType::try_new(fields)?.into();
+        let mut checkpoint_action = None;
+        let mut domain_metadata_visitor = DomainMetadataVisitor::new(None);
+        let mut set_transaction_visitor = SetTransactionVisitor::new(None);
+        for batch in snapshot.log_segment().read_actions(engine, schema)? {
+            let batch = batch?;
+            let data = batch.actions.as_ref();
+            if checkpoint_action.is_none() {
+                checkpoint_action = CheckpointAction::try_new_from_data(data)?;
+            }
```
Comment (kyli87) [inline kernel/src/transaction/root_manifest_file.rs:150]: You can early terminate here if all 3 are known by this point?

File: kernel/src/transaction/root_manifest_file.rs [hunk 23]
```diff
@@ -0,0 +1,586 @@
...
+            snapshot.table_configuration(),
+            &[],
+            &[],
+        )?;
+        assert_eq!(checkpoint.path(), manifest.file.location.as_str());
+        Ok(())
+    }
+
+    #[test]
+    fn compute_checkpoint_action_rejects_a_stale_checkpoint() -> DeltaResult<()> {
```
Comment (kyli87) [inline kernel/src/transaction/root_manifest_file.rs:350]: I'm a but confused by the intent of this test - I think it relates to the same assertion that I am confused about above though

**Review by kyli87 — COMMENTED** [review]
Code structure LGTM now, some questions about the logic / edge cases.

File: kernel/src/transaction/root_manifest_file.rs (hunk 20 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:92]: > Aren't we trying to create a new checkpointAction here, so isn't it valid that prior checkpoint would be older?

The check is checking existing checkpoint version vs what was read in read_snapshot_version (not the new commit version). 

This intent behind this check is to prevents callers of the new API from supplying an externally produced root manifest unless the previous commit was also a root manifest checkpoint.  This is because we don't have the guarantees that the external caller  knows about Delta logs on top of the root manifest and as a result to prevent against incorrect table states, we just limit this API for now. It's a check we can unblock at some later point.

 So concretely, if existing.version() != read_snapshot_version this means that a log was committed after the last root manifest checkpoint ("existing").

">" read_snapshot_version is just not possible and not exercised because the existing_checkpoint is itself bounded to the read_snapshot_version, so we can simplify the check that way.

Now, I do think the check could be made less opaque because really what we want is something like 
```
let existing_checkpoint_is_current = existing_checkpoint
    .as_ref()
    .is_none_or(|c| c.version() == read_snapshot_version);
require!(
    existing_checkpoint_is_current,
    Error::generic(format!(
        "root manifest file commit requires the immediately preceding commit to itself be a \
         root manifest checkpoint; existing checkpoint covers version {} but snapshot is at \
         {read_snapshot_version}",
        existing_checkpoint.map(|c| c.version()).unwrap_or_default()
    ))
);


```

File: kernel/src/transaction/root_manifest_file.rs (hunk 21 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:79]: That's correct. The behavior you're describing is what's happening 
domain_metadata/transactions which are keyed off of domain/appID respectively. merge_domain_metadata will override existing keys. Same idea in transactions.extend, it calls insert per pair which will replace on key collision. Finally,  since the `dm_changes` and `set_transactions` are applied after the scanned ones , we effectively guarantee that the recent one wins.

I'll work on a test though to prove this, we do have a test gap here.

Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:79]: Added a test compute_checkpoint_action_new_change_wins to confirm this behavior

Comment (kyli87) [reply kernel/src/transaction/root_manifest_file.rs:79]: Oh ok thanks, I missed that `transactions` is a `map` and not `vec`.

File: kernel/src/transaction/root_manifest_file.rs (hunk 20 above)
Comment (kyli87) [reply kernel/src/transaction/root_manifest_file.rs:92]: Hmm interesting - so you're basically saying you want every single commit to be covered by a `checkpoint` action? What about the case where there's no checkpoint action at all?

I feel like the error message could be clarified a bit:
```
require!(
                existing.version() >= read_snapshot_version,
                Error::generic(format!(
                    "root manifest file commit requires entire snapshot to be covered by a \
                     checkpoint action; existing checkpoint covers version {} but snapshot is \
                     at {read_snapshot_version}",
                    existing.version()
                ))
            );
```

**Review by kyli87 — APPROVED** [review]

File: kernel/src/transaction/root_manifest_file.rs [hunk 24]
```diff
@@ -0,0 +1,631 @@
...
+            self.scan_non_content_metadata(engine)?;
+
+        merge_domain_metadata(
+            &mut domain_metadata,
+            dm_changes
+                .iter()
+                .cloned()
+                .map(|dm| (dm.domain().to_string(), dm)),
+        );
+        transactions.extend(
```
Comment (emkornfield) [inline kernel/src/transaction/root_manifest_file.rs:71]: why does this extend and the domain metadata merge?

Comment (emkornfield) [reply kernel/src/transaction/root_manifest_file.rs:71]: oh I see this is answered below.

File: kernel/src/transaction/root_manifest_file.rs (hunk 21 above)
Comment (emkornfield) [reply kernel/src/transaction/root_manifest_file.rs:79]: maybe lets add a small comment?

File: kernel/src/transaction/root_manifest_file.rs (hunk 20 above)
Comment (emkornfield) [reply kernel/src/transaction/root_manifest_file.rs:92]: I think the initial use-case we want is this case.  We can relax/refine in a follow-up PR.

File: kernel/src/transaction/root_manifest_file.rs [hunk 25]
```diff
@@ -0,0 +1,631 @@
...
+                     at {read_snapshot_version}",
+                    existing.version()
+                ))
+            );
+        }
+
+        let version = version_as_i64(commit_version)?;
+        let content_root = ContentRoot::new(
+            self.file.location.to_string(),
+            self.file.size as i64,
```
Comment (emkornfield) [inline kernel/src/transaction/root_manifest_file.rs:97]: be careful of casts that can fail, lets make sure we don't panic here.

File: kernel/src/transaction/update.rs (hunk 15 above)
Comment (emkornfield) [reply kernel/src/transaction/update.rs:162]: bump?

**Review by emkornfield — APPROVED** [review]

File: kernel/src/transaction/root_manifest_file.rs [hunk 26]
```diff
@@ -0,0 +1,722 @@
...
+            transactions_from_crc.is_none(),
+        );
+        let mut domain_metadata = domain_metadata_from_crc
+            .unwrap_or_else(|| domain_metadata_visitor.into_domain_metadatas());
+        let mut transactions =
+            transactions_from_crc.unwrap_or(set_transaction_visitor.set_transactions);
+
+        // The log scan reads only top-level rows; fold in the checkpoint's nested set too.
+        // TODO: fold in sidecar-spilled domain metadata and transactions, not just inline entries.
+        if let Some(checkpoint) = &checkpoint_action {
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:132]: Yeah we cannot produce this state today from kernel (we don't produce spilled sidecar files yet for the new manifest checkpoint) so the premise is not valid but we can defensively verify on read in case some other writer happens to do that, so that way we don't fail to include expected DM/transaction metadata. Fixed in the latest commit.

File: kernel/src/transaction/root_manifest_file.rs [hunk 27]
```diff
@@ -0,0 +1,722 @@
...
+                     at {read_snapshot_version}",
+                    existing.version()
+                ))
+            );
+        }
+
+        let version = version_as_i64(commit_version)?;
+        let content_root = ContentRoot::new(
+            self.file.location.to_string(),
+            self.file.size as i64,
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs]: Fixed

File: kernel/src/transaction/root_manifest_file.rs [hunk 28]
```diff
@@ -0,0 +1,722 @@
...
+        let mut domain_metadata = domain_metadata_from_crc
+            .unwrap_or_else(|| domain_metadata_visitor.into_domain_metadatas());
+        let mut transactions =
+            transactions_from_crc.unwrap_or(set_transaction_visitor.set_transactions);
+
+        // The log scan reads only top-level rows; fold in the checkpoint's nested set too.
+        // TODO: fold in sidecar-spilled domain metadata and transactions, not just inline entries.
+        if let Some(checkpoint) = &checkpoint_action {
+            if from_log_scan.0 {
+                merge_domain_metadata(
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:146]: We just read the domain metadata and transactions straight from the checkpoint. We already require the last commit to be a checkpoint at the current version, so there's nothing after it we'd need to fold in or could get wrong at least with the current invariants of this API.

File: kernel/src/transaction/update.rs [hunk 29]
```diff
@@ -148,6 +154,16 @@ impl Transaction {
         self
     }
 
+    /// Stages `file` to be committed as the table's root manifest.
+    #[cfg(feature = "adaptive-metadata-in-dev")]
+    pub fn with_root_manifest_file(mut self, file: FileMeta) -> DeltaResult<Self> {
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/update.rs:253]: Intentionally removing the location check for now. We can add it later, but for now kernel will just take whatever location is produced.

File: kernel/src/actions/mod.rs [hunk 30]
```diff
@@ -1473,6 +1475,50 @@ impl CheckpointAction {
         }
     }
 
+    /// Serialize this checkpoint action into a single-row `EngineData` for `schema`.
+    #[internal_api]
+    pub(crate) fn into_engine_data(
```
Comment (amogh-jahagirdar) [inline kernel/src/actions/mod.rs:1479]: added this as part of rebasing after https://github.com/delta-io/delta-kernel-rs/pull/3285 merged

File: kernel/tests/integration/write/root_manifest_file.rs [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+        &[],
+        true,
+        READER_FEATURES.to_vec(),
+        WRITER_FEATURES.to_vec(),
+        "id",
+    )
+    .await?;
+
+    // Return the snapshot's canonicalized root (macOS resolves `/var` -> `/private/var`) so
+    // manifest paths match the locality check.
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/tests/integration/write/root_manifest_file.rs]: it was ai slop comment, removed

File: kernel/src/transaction/root_manifest_file.rs [hunk 32]
```diff
@@ -0,0 +1,819 @@
...
+            if transactions_from_crc.is_none() {
+                set_transaction_visitor.visit_rows_of(data)?;
+            }
+        }
+
+        // Reject a checkpoint that spilled txns/domain metadata to sidecars, since sidecars aren't
+        // read yet and that state would be lost.
+        if let Some(checkpoint) = &checkpoint_action {
+            require!(
+                checkpoint.txn_sidecars.is_empty()
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:134]: Added a test for this that tests both sidecar cases rather than just one

File: kernel/src/transaction/root_manifest_file.rs [hunk 33]
```diff
@@ -0,0 +1,819 @@
...
+use crate::table_configuration::TableConfiguration;
+use crate::utils::require;
+use crate::{version_as_i64, DeltaResult, Engine, FileMeta, RowVisitor as _, Version};
+
+/// A pointer to an on-disk root manifest file to be committed as the table's content root via a
+/// `checkpoint` action.
+pub(super) struct RootManifestFile {
+    pub(super) file: FileMeta,
+    /// The snapshot being updated, whose active content the checkpoint action folds in.
+    pub(super) read_snapshot: SnapshotRef,
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:22]: This is a bit more involved and ends up changing compute_checkpoint_action to be less readable imo, so I'd prefer to just keep it as is

File: kernel/src/transaction/update.rs (hunk 15 above)
Comment (amogh-jahagirdar) [reply kernel/src/transaction/update.rs]: I tightened this up, this case can only happen (as far as I can tell) if it's some kind of CTAS (e.g. doing a write where there's no previous snapshot). For now though I just removed reference to create_table, and just have a generic "Hey you can't do this if there's no previous snapshot" message rather than inferring it's always a create table.

File: kernel/src/transaction/root_manifest_file.rs [hunk 34]
```diff
@@ -0,0 +1,819 @@
...
+            table_config.protocol().clone(),
+            table_config.metadata().clone(),
+            transactions.into_values().collect(),
+            domain_metadata.into_values().collect(),
+        ))
+    }
+
+    /// Returns the read snapshot's active domain metadata, set transactions, and latest checkpoint
+    /// action.
+    fn scan_non_content_metadata(
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:101]: I went ahead and addrssed this by adding a get_app_ids to snapshot which already looked to be a todo on that anyways. It did simplify the control flow on scan_non_content_metadata, cc @kyli87 let me know what you think about this one. Since it did look like an existing TODO on snapshot I went ahead and added the function, but if you think it doesn't fit can go back to what i had before

[the remark this answers is not in the document]
Comment (kyli87) [reply kernel/src/transaction/root_manifest_file.rs:101]: I think this seems fine, the main tradeoff I see is that now we have to potentially do 3 total replays (one for each of set transaction, domain metadata, and checkpoint action) instead of getting all 3 in one. But this doesn't seem like a perf-sensitive path and most of the time it will come from CRC anyway.

**Review by kyli87 — APPROVED** [review]

**Review by emkornfield — APPROVED** [review]

File: kernel/src/transaction/root_manifest_file.rs [hunk 35]
```diff
@@ -0,0 +1,759 @@
...
```
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply kernel/src/transaction/root_manifest_file.rs:101]: I'll leave this as is. I originally had this in snapshot but folks didn't think it was a well generalized abstraction to be in snapshot, which I think is correct, because this particular scan is relevant only for root manifests checkpoints. We can always abstract that later.
